### ISSUE TITLE
ament_copyright enforces header not required by Apache 2.0

### ISSUE THREAD

**Issue by rotu** [ament/ament_lint#246]
ament_copyright is overzealous in enforcing its preferred header for Apache 2.0 license.

See http://www.apache.org/foundation/license-faq.html

* Use of the license *does not require* this header - only if it's likely that the source files somehow become separated from the license file.
* This check precludes the short variant suggested by Apache:
```
Copyright [yyyy] [name of copyright owner]
SPDX-License-Identifier: Apache-2.0
```
* Even the ASF doesn't use a copyright in the header: https://www.apache.org/legal/src-headers.html#headers

**Comment by tfoote** [comment]
@rotu Please consider rewording issues like this to be a feature request instead of an implicit criticism in that it doesn't have the feature that you want. It will go a lot further in terms of engendering good will from other developers.

This was initially implemented to meet the minimum requirements for our code base and our [developer guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/) and extra features such as supporting alternative short forms of the license were not necessary. If you would like the capability to use alternative forms of the apache license in your package that's definitely something that would be great to support.

You propose several changes here I'll go through them one by one.

Support for the short variant of the header definitely makes sense to add. 

For the copyright statement I would suggest that we continue to enforce it's presence. I took a look at the Apache documentation you linked to: https://www.apache.org/legal/src-headers.html#headers which is for their internal usage. And you're right that they don't require the copyright statement for content directly hosted by Apache projects. But they also have [various contributors agreements](https://www.apache.org/licenses/contributor-agreements.html) which means that all code in their projects is known to be copyrighted by Apache. In our case of a large ecosystem with many contributors and no CLA it's important to properly track the copyright for files. A flag could be added to specifically relax that for a file or package, but in general I think it's good to have it required by default.

I don't know exactly their justification for not putting any copyright header into the header files for Apache projects. But we definitely have header only implementations and have a standard of putting the copyright header at the top of all files so again, this is correct for our developer guide and best practices. It could be extended to support non-complying implementations. But unless we change our developer guide we should continue to enforce it is available in all headers. And if it's a lines of code issue adding the short format support would certainly help that. We have header only implementation and again run into the heterogeneous codebase with multiple authors so tracking the license of the header files is important too.

**Comment by rotu** [comment]
> Please consider rewording issues like this to be a feature request instead of an implicit criticism in that it doesn't have the feature that you want. It will go a lot further in terms of engendering good will from other developers.

I could definitely reword it to be less confrontational, but the reword you've done is not complete. This tool enforces one particular and problematic way to conform a project to the Apache license. It confuses *this header* with the requirements under the Apache license.

The job of the issue is to document the issue. The fix PR(s) are where a particular solution should be chosen and implemented.

> But they also have various contributors agreements which means that all code in their projects is known to be copyrighted by Apache.

That it not true. Under the Apache license, and the CLA, the contributions are permissively licensed to the licensor - the copyright is not *assigned* to the current copyright holder.

The expected way to include additional notices of authorship at ASF is to move them from the source code to the NOTICE file. https://www.apache.org/legal/src-headers.html#headers

> In our case of a large ecosystem with many contributors and no CLA it's important to properly track the copyright for files.

We don't do that. The CLA doesn't track who the copyright holder is, i.e. my employer. Barring an additional copyright header, there's not a good way to do that.

> A flag could be added to specifically relax that for a file or package, but in general I think it's good to have it required by default.

👍

> I don't know exactly their justification for not putting any copyright header into the header files for Apache projects.

I assume because the copyright holder is not usually important to the reader of the file and, in open source, usually there are many of them. CLA's acknowledge that: "we don't care who owns this code so long as you vouch that we can use it"

Here are the justifications as I see it. There are two things here, the copyright and the license that need to be considered:

Having the license at the top of each file:
* unnecessary: it is not *required* to license the file; it only prevents accidentally separating the license from individual source files.
* clutter: this takes up space
* contradictory: if this contradicts the project-level license, accidentally or maliciously, that's a problem.

Having the copyright at the top of each file:
* unnecessary: copyright protection does not require a copyright notice
* clutter: keeping the notice current makes it arbitrarily long if it is updated on each commit.
* contradictory: usually the copyright is *not* updated on each contribution. This means the license file and/or the license header falsely imply there is a single copyright holder

> this is correct for our developer guide and best practices.
>...
> But unless we change our developer guide we should continue to enforce it is available in all headers.

True: the docs say to use the automated linters. I just chose to confront it in the linters first.

Best practice, according Apache, is to move all legal notices possible to the NOTICE file. https://www.apache.org/dev/licensing-howto.html
